apis/projectcontour/v1alpha1/accesslog.go

```diff
@@ -0,0 +1,251 @@
+var envoySimpleOperators = map[string]struct{}{
+	"BYTES_RECEIVED":                                {},
+	"BYTES_SENT":                                    {},
+	"CONNECTION_ID":                                 {},
+	"CONNECTION_TERMINATION_DETAILS":                {},
+	"DOWNSTREAM_DIRECT_REMOTE_ADDRESS":              {},
+	"DOWNSTREAM_DIRECT_REMOTE_ADDRESS_WITHOUT_PORT": {},
+	"DOWNSTREAM_LOCAL_ADDRESS":                      {},
+	"DOWNSTREAM_LOCAL_ADDRESS_WITHOUT_PORT":         {},
+	"DOWNSTREAM_LOCAL_PORT":                         {},
+	"DOWNSTREAM_LOCAL_SUBJECT":                      {},
+	"DOWNSTREAM_LOCAL_URI_SAN":                      {},
+	"DOWNSTREAM_PEER_CERT":                          {},
+	"DOWNSTREAM_PEER_CERT_V_END":                    {},
+	"DOWNSTREAM_PEER_CERT_V_START":                  {},
+	"DOWNSTREAM_PEER_FINGERPRINT_1":                 {},
+	"DOWNSTREAM_PEER_FINGERPRINT_256":               {},
+	"DOWNSTREAM_PEER_ISSUER":                        {},
+	"DOWNSTREAM_PEER_SERIAL":                        {},
+	"DOWNSTREAM_PEER_SUBJECT":                       {},
+	"DOWNSTREAM_PEER_URI_SAN":                       {},
+	"DOWNSTREAM_REMOTE_ADDRESS":                     {},
+	"DOWNSTREAM_REMOTE_ADDRESS_WITHOUT_PORT":        {},
+	"DOWNSTREAM_TLS_CIPHER":                         {},
+	"DOWNSTREAM_TLS_SESSION_ID":                     {},
+	"DOWNSTREAM_TLS_VERSION":                        {},
+	"DURATION":                                      {},
+	"GRPC_STATUS":                                   {},
+	"HOSTNAME":                                      {},
+	"LOCAL_REPLY_BODY":                              {},
+	"PROTOCOL":                                      {},
+	"REQUESTED_SERVER_NAME":                         {},
+	"REQUEST_DURATION":                              {},
+	"RESPONSE_CODE":                                 {},
+	"RESPONSE_CODE_DETAILS":                         {},
+	"RESPONSE_DURATION":                             {},
+	"RESPONSE_FLAGS":                                {},
+	"RESPONSE_TX_DURATION":                          {},
+	"ROUTE_NAME":                                    {},
+	"START_TIME":                                    {},
+	"UPSTREAM_CLUSTER":                              {},
+	"UPSTREAM_HOST":                                 {},
+	"UPSTREAM_LOCAL_ADDRESS":                        {},
+	"UPSTREAM_TRANSPORT_FAILURE_REASON":             {},
+}
+
+// envoyComplexOperators is the list of known Envoy log template keywords that require
+// arguments.
+var envoyComplexOperators = map[string]struct{}{
+	"REQ":               {},
+	"RESP":              {},
+	"START_TIME":        {},
+	"TRAILER":           {},
+	"REQ_WITHOUT_QUERY": {},
+}
+
+// AccessLogType is the name of a supported access logging mechanism.
+type AccessLogType string
+
+func (a AccessLogType) Validate() error {
+	switch a {
+	case EnvoyAccessLog, JSONAccessLog:
+		return nil
+	default:
+		return fmt.Errorf("invalid access log format %q", a)
+	}
+}
```

Comment:
Would it be worth moving these validation helpers into an internal package so that we didn't have to deal with API compatibility guarantees? I'd sort of prefer that all else being equal.

Related, wondering if there's a way to reuse implementations across here and `pkg/config` so we don't have to maintain both. I know the `pkg/config` ones will go away eventually, but probably not until 2.0 since they're in a public package. Perhaps we could internally convert the config file structs into the CRD structs (or vice versa) and then just have one validation implementation. Would have to poke around a little more to see exactly what that would look like.

Reply:
yeah I agree and brought that up before but the team at the time felt we should have them in the public API. 

Some of the issue with the `pkg/config` resuse is that the names not being camel-cases and other groupings. 

For validation, we may need to convert Contour internally to use the CRD configuration spec and deprecate out the `ServeContext`, having a converter migrate the `ServeContext` into a `ContourConfiguration`.

Reply:
I pushed for these to be in here so that anyone who consumes HTTPProxy or any Contour CRDs would have the same validation logic available to them, which makes actually reusing them in other cases more likely to be consistent.

I think that we should absolutely be using only one representation internally inside Contour, and ideally it should be these new structs (since this is where we are heading). But I'm not sure if we do that now or later. In either case, we're going to end up doing work that will end up thrown away translating from one thing to another, I think that the decision is more around how much do we do before 1.19 in the initial release vs later as refactoring.

Reply:
In this specific case, using a kubebuilder validation is actually heaps better, since that will mean that the object can't be applied unless the validation passes. Kubebuilder has enums available already too.

Reply:
Just need to be careful about some of that. IngressClass is a great example. If it's not defined, Contour looks at ALL ingress objects, if it's defaulted to `Contour` then it won't process any resource without a class defined.

Reply:
> I think that we should absolutely be using only one representation internally inside Contour, and ideally it should be these new structs (since this is where we are heading). But I'm not sure if we do that now or later. In either case, we're going to end up doing work that will end up thrown away translating from one thing to another, I think that the decision is more around how much do we do before 1.19 in the initial release vs later as refactoring.

I agree, we can definitely revisit this after the initial release. Ultimately we might decide it's not worth it given that we'll throw the old stuff away.